oauth2client/contrib/multiprocess_file_storage.py

```diff
@@ -0,0 +1,247 @@
+# Copyright 2016 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""File-based storage that supports multiple credentials and cross-process
+access.
+
+This module implements a credential store where multiple
+credentials are stored in one file. That file supports locking between threads
+and processes.
+
+This module superceeds the functionality previously found in `multistore_file`.
+
+The credential themselves are keyed off of:
+
+* client_id
+* user_agent
+* scope
+
+The format of the stored data is like so::
+
+    {
+      'file_version': 2,
+      'credentials': {
+        'key': 'base64 credentials'
+      }
+    }
+
+"""
+
+import base64
+import json
+import logging
+import os
+import threading
+
+import fasteners
+from six import iteritems
+
+from oauth2client.client import Credentials
+from oauth2client.client import Storage as BaseStorage
+from oauth2client import util
+
+logger = logging.getLogger(__name__)
+_backends = {}
+_backends_lock = threading.Lock()
+
+
+@util.positional(1)
+def _get_backend(filename):
+    """A helper method to get or create a backend with thread locking.
+
+    There should only be one backend per-file per-process."""
+    if filename in _backends:
+        return _backends[filename]
```

Comment:
This is an unlocked read. LBYL vs. EAFP matters when race conditions might occur. If you want to avoid locking:

```python
try:
    return _backends[filename]
except KeyError:
    pass
```